src/main/java/org/truffleruby/core/regexp/MatchDataNodes.java

```diff
@@ -282,14 +281,59 @@ protected boolean inBounds(DynamicObject matchData, int index) {
     }
 
 
+    public abstract static class ValuesNode extends CoreMethodArrayArgumentsNode {
+
+        @Child private RopeNodes.MakeSubstringNode makeSubstringNode = RopeNodes.MakeSubstringNode.create();
+
+        public static ValuesNode create() {
+            return ValuesNodeFactory.create(new RubyNode[0]);
+        }
+
+        public abstract Object[] execute(DynamicObject matchData);
+
+        @Specialization(guards = "hasValues(matchData)")
+        public Object[] getValuesFast(DynamicObject matchData) {
+            return Layouts.MATCH_DATA.getValues(matchData);
+        }
+
+        @TruffleBoundary
+        @Specialization(guards = "!hasValues(matchData)")
+        public Object[] getValuesSlow(DynamicObject matchData) {
+            DynamicObject source = Layouts.MATCH_DATA.getSource(matchData);
```

Comment:
It's not required, but we generally make locals `final` and you seem to have a mix of that going on here.

Reply:
It occurs to me that this was probably imported JRuby code. Nevermind.

Reply:
They were pretty inconsistent in the old `matchCommon` and I should have fixed that when I pulled it apart.